cpmpy/solvers/gurobi.py

```diff
@@ -265,12 +265,12 @@ def transform(self, cpm_expr):
         # apply transformations, then post internally
         # expressions have to be linearized to fit in MIP model. See /transformations/linearize
         cpm_cons = flatten_constraint(cpm_expr)  # flat normal form
-        cpm_cons = decompose_global(cpm_cons, supported={"min","max", "element", "alldifferent"}) #element and alldiff have specialized MIP decomp in linearize
+        cpm_cons = decompose_global(cpm_cons, supported={"min","max","alldifferent"}) # alldiff has specialized MIP decomp in linearize
         cpm_cons = reify_rewrite(cpm_cons, supported=frozenset(['sum', 'wsum']))  # constraints that support reification
         cpm_cons = only_bv_implies(cpm_cons)  # anything that can create full reif should go above...
-        cpm_cons = linearize_constraint(cpm_cons)  # the core of the MIP-linearization
         cpm_cons = only_numexpr_equality(cpm_cons, supported=frozenset(["sum", "wsum", "sub"]))  # supports >, <, !=
-        cpm_cons = only_positive_bv(cpm_cons)  # after linearisation, rewrite ~bv into 1-bv
+        cpm_cons = linearize_constraint(cpm_cons, supported=frozenset({"sum", "wsum","sub","min","max","mul","abs","pow","div"}))  # the core of the MIP-linearization
+        cpm_cons = only_positive_bv(cpm_cons)  # after linearization, rewrite ~bv into 1-bv
```

Comment:
Can we have `5*~bv` as input here, so that it is transformed to `5*(1-bv)`? How do we avoid introducing nested expressions here?

Reply:
Anything that comes out of linearize is either a wsum, sum or some supported expression/global constraint.
For sum and wsum, it is trivial to put all of the constants on the rhs while eliminating `~bv`.
For the other expressions, we loop over the arguments in the expression and replace `~bv` with a new variable `bvn` and add the constraint `bvn == 1-bv`.
Aha but we still have to linearize this new constraint... will at that!
For global constraints we do the same thing, although I haven't found a MIP solver yet supporting global constraints so it hasn't come up yet :)

Reply:
Hmm, it is correct, but I don't quite like the `bvn == 1-bv`. That's introducing a new variable for no good reason... 

So we get `3*~x >= ~y` to become `[3*xn >= yn, xn + x == 1, yn + y == 1]`, which is 5 constraints in Exact instead of just one: `y-3x >= 2`. 

The natural solution would be to have `3*~x >= ~y` become `3*(1-x) >= 1-y`, apply distributivity, then normalize during linearization. This is tricky, because we introduce nesting where we want to keep things flattened.

This would actually be doable if everything was a weighted sum: `ws([3],[~x]) >= ws([1],[~y])` can easily be transformed to `ws([3,-3],[1,x]) >= ws([1,-1],[1,y])` after which normalization (moving all constants to the right and all non-constants to the left) fixes the problem.

So here's my proposal: turn every multiplication/regular sum to a weighted sum, run `only_positive_bv` then `linearize_constraint`.

Alternatively, make `only_positive_bv` generate weighted sums, split the normalization from linearization, linearize -> only_positive_bv -> normalize.

Reply:
The distributivity trick is exactly what happens in the current pipeline:
`3 * ~x >= ~y` will be linearized to `[sum([3, -1] * [~x, ~y]) >= 0]`. After which `only_positive_bv` will transform it to `[sum((-3, 1) * (x, y)) >= -2]`.

The case I was talking about for when we take out the `~bv` is for example in `max(bv1, ~bv2) <= 1`. This constraint will be kept as is by linearize in case `max` is in the supported list.
Then `only_positive_bv` will transform it to `[max(bv1, aux) <= 1, bv2 + aux = 1]`